### PR TITLE
MVP2-402 - Number scaling changes

### PR DESCRIPTION
**Pull request by SamirHodzic**
@rudolfix if you are able to test and compare, it would be awesome. 
On places which I was able to track down, numbers changed a bit, and outcome calculation numbers are kinda better.

### CODE REVIEW

File: controllers/users-controller.js [hunk 1]
```diff
@@ -282,9 +282,9 @@ const getAMMHistory = async (req, res, next) => {
       const transactions = [];
 
       for (const interaction of interactions) {
-        const investmentAmount = toPrettyBigDecimal(BigInt(interaction.investmentamount));
-        const feeAmount = toPrettyBigDecimal(BigInt(interaction.feeamount));
-        const outcomeTokensBought = toPrettyBigDecimal(BigInt(interaction.outcometokensbought));
+        const investmentAmount = toBigDecimal(BigInt(interaction.investmentamount));
```
Comment (rudolfix) [inline controllers/users-controller.js:285]: what about `getTransactions` function above? it returns big int data directly. is it used at all? IMO data should be converted like here

File: controllers/users-controller.js [hunk 2]
```diff
@@ -334,7 +334,7 @@ const getTradeHistory = async (req, res, next) => {
         const totalSellAmount = _.sum(
```
Comment (rudolfix) [inline controllers/users-controller.js:334]: this is super fishy. are we converting there to Number to add all the elements? can't we just add BigInts dirctly? no floats please

File: controllers/users-controller.js [hunk 3]
```diff
@@ -334,7 +334,7 @@ const getTradeHistory = async (req, res, next) => {
...
           sellInteractions.map(_.property('investmentamount')).map(Number).filter(_.isFinite)
         );
-        soldAmount = toPrettyBigDecimal(BigInt(totalSellAmount || 0));
+        soldAmount = toBigDecimal(BigInt(totalSellAmount || 0n));
```
Comment (rudolfix) [inline controllers/users-controller.js:337]: then this double conversion does not make sense. and why you do `|| 0n`? you are ORing BigNumber with BigInt here. that cannot work

**Review by rudolfix — CHANGES_REQUESTED** [review]
also please fix the following

```
exports.provideLiquidityToBet = async (createBet) => {
  const LOG_TAG = '[CREATE-BET]';
  const liquidityAmount = 100_0000n; // bets start with 100 liquidity
```
change it to 10000 and provide a valid comment:

in admin-service there's a huge code duplication:
```
 'do-resolve': {
              // create a totally new action
              actionType: 'record',
              isVisible: false,
              handler: async (request, response, context) => {
                const id = context.record.params._id;
                const bet = await eventService.getBet(id);
                const event = await Event.findById(bet.event);
                const indexOutcome = request.fields.index;
                const { evidenceActual } = request.fields;
                const { evidenceDescription } = request.fields;

                if (bet.status !== 'active' && bet.status !== 'closed') {
                  context.record.params.message =
                    'Event can only be resolved if it is active or closed';
                  return {
                    record: context.record.toJSON(context.currentAdmin),
                  };
                }

                let resolveResults = [];
                let ammInteraction = [];

                const session = await Bet.startSession();
                try {
                  await session.withTransaction(async () => {
                    bet.finalOutcome = indexOutcome;
                    bet.resolved = true;
                    bet.evidenceDescription = evidenceDescription;
                    bet.evidenceActual = evidenceActual;

                    await betService.clearOpenBets(bet, session);
                    await bet.save({ session });
                    const betContract = new BetContract(id);
                    resolveResults = await betContract.resolveAndPayout(
                      'Wallfair Admin User',
                      indexOutcome
                    );
                    ammInteraction = await betContract.getUserAmmInteractions();
                  });
                } catch (err) {
                  console.debug(err);
                } finally {
                  await session.endSession();

                  // find out how much each individual user invested
                  const investedValues = {}; // userId -> value
                  for (const interaction of ammInteraction) {
                    const amount = Number(interaction.amount) / Number(WFAIR.ONE);
                    if (interaction.direction === 'BUY') {
                      // when user bought, add this amount to value invested
                      investedValues[interaction.buyer] = investedValues[interaction.buyer]
                        ? investedValues[interaction.buyer] + amount
                        : amount;
                    } else if (interaction.direction === 'SELL') {
                      // when user sells, decrease amount invested
                      investedValues[interaction.buyer] =
                        investedValues[interaction.buyer] - amount;
                    }
                  }

                  for (const resolvedResult of resolveResults) {
                    const userId = resolvedResult.owner;
                    const { balance } = resolvedResult;

                    const winToken = Math.round(Number(balance) / Number(WFAIR.ONE));

                    if (userId.includes('_')) {
                      continue;
                    }

                    // update the balance of tokens won of a user, to be used for leaderboards
                    // must be done inside transaction
                    await userService.increaseAmountWon(userId, winToken);

                    // send notification to this user
                    websocketService.emitBetResolveNotification(
                      userId,
                      id,
                      bet.marketQuestion,
                      bet.outcomes[indexOutcome].name,
                      Math.round(investedValues[userId]),
                      event.previewImageUrl,
                      winToken
                    );
                    
```

is it being still used? then fix the scaling there - as fixed in the other place. or merge the code

File: controllers/users-controller.js (hunk 1 above)
Comment (SamirHodzic) [reply controllers/users-controller.js:285]: Removed that endpoint, not used anymore.

File: services/bet-service.js [hunk 4]
```diff
@@ -287,8 +283,7 @@ exports.resolve = async ({
   for (const resolvedResult of resolveResults) {
     const userId = resolvedResult.owner;
     const { balance } = resolvedResult;
-
-    const winToken = Math.round(Number(balance) / Number(WFAIR.ONE));
+    const winToken = toBigDecimal(balance);
```
[the remark this answers is not in the document]
Comment (SamirHodzic) [reply services/bet-service.js:286]: We are removing this adminbro functions today, so not needed.

File: controllers/users-controller.js [hunk 5]
```diff
@@ -332,9 +314,9 @@ const getTradeHistory = async (req, res, next) => {
             i.outcome === bet.outcomeIndex
         );
         const totalSellAmount = _.sum(
-          sellInteractions.map(_.property('investmentamount')).map(Number).filter(_.isFinite)
+          sellInteractions.map(_.property('investmentamount')).map(BigInt).filter(Boolean)
```
Comment (rudolfix) [inline controllers/users-controller.js:317]: I just trust that it works :)

File: services/bet-service.js [hunk 6]
```diff
@@ -313,7 +307,7 @@ exports.resolve = async ({
       betId,
       bet.marketQuestion,
       bet.outcomes[outcomeIndex].name,
-      Math.round(investedValues[userId]),
+      +investedValues[userId],
```
Comment (rudolfix) [inline services/bet-service.js:310]: was is this "+" doing here?

**Review by rudolfix — APPROVED** [review]

File: controllers/users-controller.js (hunk 5 above)
Comment (SamirHodzic) [reply controllers/users-controller.js:317]: It sums all `investmentamount` firstly mapped from string to BigInt and filtering for any kind of not-number values.

File: services/bet-service.js (hunk 6 above)
Comment (SamirHodzic) [reply services/bet-service.js:310]: Just to convert string value `"1234.5500" -> 1234.5500` for comparison while sending socket message to user (if value larger than zero, there is additional message part provided).

**Pull request merged**